### ISSUE TITLE
Cannot iterate over the same table

### ISSUE THREAD

**Issue by nilium** [Shopify/go-lua#81]
This is a side-effect of #80 being reported, but because iteration over tables causes the table's iterationKeys field to be assigned, it is not currently possible to perform nested iteration over the same table. For example, the following test will fail (but it will pass in regular Lua 5.2):

```go
func TestConcurrentNext(t *testing.T) {
	testString(t, `
	t = {}
	t[1], t[2], t[3] = true, true, true

	r = {}
	for k1 in pairs(t) do
		r[#r + 1] = k1
		for k2 in pairs(t) do
			r[#r + 1] = k2
		end
	end

	got = table.concat(r, "")
	want = "112321233123"
	assert(got == want, "got " .. got .. "; want " .. want)
	`)
}
```

**Comment by nilium** [comment]
I've implemented a fix for this and #80, but it would constitute a leak since we'd need to keep dead keys around to satisfy how `next` works in Lua. I think this necessitates actually implementing Lua's GC, since further work on this would end up diverging probably more than I'd like. If anyone has suggestions on where to go with this, I'd appreciate it, since I think I'm talking myself into a corner on how to approach the two issues.

**Comment by fbogsany** [comment]
The test for this is invalid because it makes assumptions about the iteration order. The documentation for `next()` states (emphasis in original):
> The order in which the indices are enumerated is not specified, _even for numeric indices_. (To traverse a table in numerical order, use a numerical **for**.)

A more reasonable test is:
```go
func TestConcurrentNext(t *testing.T) {
	testString(t, `
	t = {}
	t[1], t[2], t[3] = true, true, true

	outer = {}
	for k1 in pairs(t) do
		table.insert(outer, k1)
		inner = {}
		for k2 in pairs(t) do
			table.insert(inner, k2)
		end
		table.sort(inner)
		got = table.concat(inner, '')
		assert(got == '123', 'got ' .. got .. '; want 123')
	end

	table.sort(outer)
	got = table.concat(outer, '')
	assert(got == '123', 'got ' .. got .. '; want 123')
	`)
}
```
